drivers/sensor/bmm150/bmm150.c

```diff
@@ -0,0 +1,595 @@
+                                                &max_odr) < 0) {
+                        return -EIO;
+                }
+
+                if (data->odr <= 0) {
+                        if (bmm150_read_odr(dev) < 0) {
+                                return -EIO;
+                        }
+                }
+
+                if (data->odr > max_odr) {
+                        return -EINVAL;
+		}
+
+		if (bmm150_write_rep_z(dev, val->val1) < 0) {
+                        return -EIO;
+                }
+                break;
+#endif
+        default:
+                return -EINVAL;
+        }
+
+        return 0;
+}
+#endif
+
+#if defined(BMM150_SET_ATTR)
+static int bmm150_attr_set(struct device *dev,
+                                enum sensor_channel chan,
+                                enum sensor_attribute attr,
+                                const struct sensor_value *val)
+{
+        struct bmm150_magn_data *data = dev->driver_data;
+
+        switch (attr) {
+#if defined(CONFIG_BMM150_SAMPLING_RATE_RUNTIME)
+        case SENSOR_ATTR_SAMPLING_FREQUENCY:
+                if (data->max_odr <= 0) {
+                        if (bmm150_compute_max_odr(dev, 0, 0,
+                                                        &data->max_odr) < 0) {
+                                return -EIO;
+                        }
+                }
+
+                if (data->max_odr < val->val1) {
+                        SYS_LOG_DBG("not supported with current oversampling");
+                        return -ENOTSUP;
+                }
+
+		if (bmm150_set_odr(dev, (u8_t)(val->val1)) < 0) {
+                        return -EIO;
+                }
+                break;
+#endif
+#if defined(BMM150_SET_ATTR_REP)
+        case SENSOR_ATTR_OVERSAMPLING:
+                bmm150_attr_set_rep(dev, chan, val);
+                break;
+#endif
+        default:
+                return -EINVAL;
+        }
+
+        return 0;
+}
+#endif
+
+static const struct sensor_driver_api bmm150_api_funcs = {
+#if defined(BMM150_SET_ATTR)
+        .attr_set = bmm150_attr_set,
+#endif
+        .sample_fetch = bmm150_sample_fetch,
+        .channel_get = bmm150_channel_get,
+};
+
+static int bmm150_init_chip(struct device *dev)
+{
+        struct bmm150_data *data = dev->driver_data;
+        const struct bmm150_config *config = dev->config->config_info;
+        u8_t chip_id;
+        struct bmm150_preset preset;
+
+        bmm150_set_power_mode(dev, BMM150_POWER_MODE_NORMAL, 0);
+        bmm150_set_power_mode(dev, BMM150_POWER_MODE_SUSPEND, 1);
+
+        if (bmm150_set_power_mode(dev, BMM150_POWER_MODE_SUSPEND, 0)
+                                       < 0) {
+                SYS_LOG_DBG("failed to bring up device from suspend mode");
+                return -EIO;
+        }
+
+        if (i2c_reg_read_byte(data->i2c, config->i2c_slave_addr,
+                              BMM150_REG_CHIP_ID, &chip_id) < 0) {
```

Comment:
According to Datasheet section 5.3 
This number can only be read if the power control bit (register 0x4B bit0) is enabled.